File: p5-api-reverse-engineering/scripts/test-upload.py
#!/usr/bin/env python3
"""
p5.js Web Editor API - Sketch Upload Test Script (Python)

PURPOSE: Test uploading sketches to editor.p5js.org programmatically using Python

REQUIREMENTS:
- Python 3.7+
- pip install requests

SETUP:
1. Log into editor.p5js.org in your browser
2. Open DevTools → Application → Cookies
3. Copy the value of 'connect.sid' cookie
4. Paste it in the SESSION_COOKIE variable below

OR (if Personal Access Tokens are working):
1. Generate token in Account Settings
2. Set USERNAME and ACCESS_TOKEN variables

USAGE:
python test-upload.py
"""

import requests
import json
import base64
import sys
from typing import Dict, List, Optional
from datetime import datetime

# ============================================================================
# CONFIGURATION - UPDATE THESE VALUES
# ============================================================================

# Method 1: Session Cookie (copy from browser DevTools)
SESSION_COOKIE = "YOUR_SESSION_COOKIE_HERE"

# Method 2: Personal Access Token (if available)
USERNAME = "your_username"
ACCESS_TOKEN = "your_personal_access_token"

# Which authentication method to use
USE_TOKEN_AUTH = False  # Set to True to use token, False for cookie

# ============================================================================
# API CLIENT CONFIGURATION
# ============================================================================

API_BASE_URL = "https://editor.p5js.org/api"


class P5APIClient:
    """Client for interacting with p5.js Web Editor API"""

    def __init__(self, session_cookie: Optional[str] = None,
                 username: Optional[str] = None,
                 access_token: Optional[str] = None):
        self.base_url = API_BASE_URL
        self.session = requests.Session()
        self.session.headers.update({
            'Content-Type': 'application/json',
            'User-Agent': 'p5-api-test-client-python/1.0'
        })

        # Setup authentication
        if access_token and username:
            # Method 1: HTTP Basic Auth with Personal Access Token
            auth_string = f"{username}:{access_token}"
            auth_bytes = auth_string.encode('utf-8')
            auth_b64 = base64.b64encode(auth_bytes).decode('utf-8')
            self.session.headers['Authorization'] = f'Basic {auth_b64}'
            print("🔑 Using Token Authentication")
        elif session_cookie:
            # Method 2: Session Cookie
            self.session.cookies.set('connect.sid', session_cookie,
                                    domain='editor.p5js.org')
            print("🔑 Using Session Cookie Authentication")
        else:
            print("⚠️  No authentication configured")

    def _request(self, method: str, endpoint: str, **kwargs) -> requests.Response:
        """Make HTTP request with logging"""
        url = f"{self.base_url}{endpoint}"
        print(f"\n📤 {method.upper()} {endpoint}")

        try:
            response = self.session.request(method, url, **kwargs)
            print(f"✅ {response.status_code} {response.reason}")
            return response
        except requests.exceptions.RequestException as e:
            print(f"❌ Request failed: {e}")
            raise

    def test_authentication(self) -> Optional[bool]:
        """Test authentication by checking access"""
        print("\n🔍 Testing Authentication...")
        try:
            response = self._request('GET', '/auth/access-check')
            if response.status_code == 200:
                print("✅ Authentication successful!")
                return True
            else:
                print("❌ Authentication failed")
                return False
        except requests.exceptions.HTTPError as e:
            if e.response.status_code == 404:
                print("⚠️  Test endpoint not found (this is okay)")
                return None
            print("❌ Authentication failed")
            return False

    def create_sketch(self, sketch_data: Dict) -> Dict:
        """Create a new sketch"""
        print(f'\n📝 Creating sketch: "{sketch_data["name"]}"...')

        response = self._request('POST', '/projects', json=sketch_data)
        response.raise_for_status()

        project = response.json()
        print("✅ Sketch created successfully!")
        print("\n📊 Project Details:")
        print(f"  ID: {project.get('id')}")
        print(f"  Name: {project.get('name')}")

        owner = project.get('owner') or project.get('user', {})
        username = owner.get('username') if isinstance(owner, dict) else str(owner)
        print(f"  Owner: {username}")
        print(f"  Created: {project.get('createdAt')}")

        sketch_url = f"https://editor.p5js.org/{username}/sketches/{project['id']}"
        print("\n🔗 Sketch URL:")
        print(f"   {sketch_url}")
        print("\n🎨 Full View:")
        print(f"   https://editor.p5js.org/{username}/full/{project['id']}")

        return project

    def get_project(self, project_id: str) -> Dict:
        """Get project details"""
        print(f"\n🔍 Fetching project {project_id}...")

        response = self._request('GET', f'/projects/{project_id}')
        response.raise_for_status()

        print("✅ Project fetched successfully")
        return response.json()

    def update_sketch(self, project_id: str, updates: Dict) -> Dict:
        """Update a sketch"""
        print(f"\n✏️  Updating sketch {project_id}...")

        response = self._request('PUT', f'/projects/{project_id}', json=updates)
        response.raise_for_status()

        print("✅ Sketch updated successfully")
        return response.json()

    def delete_sketch(self, project_id: str) -> bool:
        """Delete a sketch"""
        print(f"\n🗑️  Deleting sketch {project_id}...")

        response = self._request('DELETE', f'/projects/{project_id}')
        response.raise_for_status()

        print("✅ Sketch deleted successfully")
        return True

    def list_user_sketches(self, username: str) -> List[Dict]:
        """List user's sketches (public endpoint, no auth needed)"""
        print(f"\n📋 Listing sketches for user: {username}...")

        # This endpoint doesn't require authentication
        response = requests.get(f"{self.base_url}/{username}/sketches")
        response.raise_for_status()

        sketches = response.json()
        print(f"✅ Found {len(sketches)} sketches")

        for i, sketch in enumerate(sketches[:5], 1):
            print(f"\n  {i}. {sketch['name']}")
            print(f"     ID: {sketch['id']}")
            print(f"     Updated: {sketch['updatedAt']}")

        if len(sketches) > 5:
            print(f"\n  ... and {len(sketches) - 5} more")

        return sketches


# ============================================================================
# SKETCH TEMPLATES
# ============================================================================

TEMPLATE_BASIC = {
    "name": "Test Sketch - Basic (Python)",
    "files": [
        {
            "name": "sketch.js",
            "content": """function setup() {
  createCanvas(400, 400);
}

function draw() {
  background(220);

  // Draw a circle that follows the mouse
  fill(255, 0, 0);
  circle(mouseX, mouseY, 50);

  // Display coordinates
  fill(0);
  textAlign(CENTER);
  text(`x: ${mouseX}, y: ${mouseY}`, width / 2, 20);
}""",
            "fileType": "file"
        },
        {
            "name": "index.html",
            "content": """<!DOCTYPE html>
<html lang="en">
  <head>
    <meta charset="utf-8">
    <meta name="viewport" content="width=device-width, initial-scale=1.0">
    <title>Test Sketch</title>
    <style>
      body {
        padding: 0;
        margin: 0;
      }
    </style>
    <script src="https://cdnjs.cloudflare.com/ajax/libs/p5.js/1.4.0/p5.js"></script>
    <script src="sketch.js"></script>
  </head>
  <body>
  </body>
</html>""",
            "fileType": "file"
        }
    ]
}

TEMPLATE_ANIMATION = {
    "name": "Test Sketch - Animation (Python)",
    "files": [
        {
            "name": "sketch.js",
            "content": """let angle = 0;

function setup() {
  createCanvas(400, 400);
}

function draw() {
  background(220);

  translate(width / 2, height / 2);
  rotate(angle);

  fill(255, 0, 0);
  rectMode(CENTER);
  rect(0, 0, 100, 100);

  angle += 0.01;
}""",
            "fileType": "file"
        },
        {
            "name": "index.html",
            "content": """<!DOCTYPE html>
<html lang="en">
  <head>
    <meta charset="utf-8">
    <meta name="viewport" content="width=device-width, initial-scale=1.0">
    <title>Animation Test</title>
    <style>
      body {
        padding: 0;
        margin: 0;
      }
    </style>
    <script src="https://cdnjs.cloudflare.com/ajax/libs/p5.js/1.4.0/p5.js"></script>
    <script src="sketch.js"></script>
  </head>
  <body>
  </body>
</html>""",
            "fileType": "file"
        }
    ]
}


# ============================================================================
# TEST SCENARIOS
# ============================================================================

def run_tests(client: P5APIClient):
    """Run all tests"""
    print("╔════════════════════════════════════════════════════════════════╗")
    print("║    p5.js Web Editor API - Upload Test Suite (Python)          ║")
    print("╚════════════════════════════════════════════════════════════════╝")

    try:
        # Test 1: Authentication
        client.test_authentication()

        # Test 2: Create a basic sketch
        project = client.create_sketch(TEMPLATE_BASIC)

        # Test 3: Fetch the created project
        client.get_project(project['id'])

        # Test 4: Update the sketch
        updates = {
            "name": "Test Sketch - Updated (Python)",
            "files": TEMPLATE_ANIMATION["files"]
        }
        client.update_sketch(project['id'], updates)

        # Test 5: List user's sketches
        if USERNAME != "your_username":
            client.list_user_sketches(USERNAME)

        # Test 6: Clean up (optional - comment out if you want to keep the sketch)
        # client.delete_sketch(project['id'])

        print("\n╔════════════════════════════════════════════════════════════════╗")
        print("║    ✅ All tests completed successfully!                        ║")
        print("╚════════════════════════════════════════════════════════════════╝")

    except Exception as e:
        print("\n╔════════════════════════════════════════════════════════════════╗")
        print("║    ❌ Tests failed                                             ║")
        print("╚════════════════════════════════════════════════════════════════╝")
        print(f"\nError details: {e}")
        sys.exit(1)


# ============================================================================
# VALIDATION
# ============================================================================

def validate_config():
    """Validate configuration"""
    print("🔧 Validating configuration...")

    if not USE_TOKEN_AUTH and SESSION_COOKIE == "YOUR_SESSION_COOKIE_HERE":
        print("\n❌ ERROR: SESSION_COOKIE not configured!")
        print("\n📖 Instructions:")
        print("1. Log into editor.p5js.org in your browser")
        print("2. Open DevTools (F12) → Application → Cookies")
        print('3. Copy the value of "connect.sid" cookie')
        print("4. Paste it in this script as SESSION_COOKIE")
        print("\nOR set USE_TOKEN_AUTH = True and configure token credentials\n")
        sys.exit(1)

    if USE_TOKEN_AUTH and (USERNAME == "your_username" or ACCESS_TOKEN == "your_personal_access_token"):
        print("\n❌ ERROR: Token credentials not configured!")
        print("\n📖 Instructions:")
        print("1. Log into editor.p5js.org")
        print("2. Go to Account Settings")
        print("3. Generate a Personal Access Token")
        print("4. Set USERNAME and ACCESS_TOKEN in this script")
        print("\nOR set USE_TOKEN_AUTH = False and use session cookie instead\n")
        sys.exit(1)

    print("✅ Configuration valid\n")


# ============================================================================
# MAIN
# ============================================================================

def main():
    """Main entry point"""
    # Check if requests is installed
    try:
        import requests
    except ImportError:
        print("❌ requests library is not installed!")
        print("Run: pip install requests")
        sys.exit(1)

    validate_config()

    # Create API client with appropriate auth
    if USE_TOKEN_AUTH:
        client = P5APIClient(username=USERNAME, access_token=ACCESS_TOKEN)
    else:
        client = P5APIClient(session_cookie=SESSION_COOKIE)

    run_tests(client)


if __name__ == "__main__":
    main()

File: p5-api-reverse-engineering/scripts/test-upload.js
#!/usr/bin/env node

/**
 * p5.js Web Editor API - Sketch Upload Test Script
 *
 * PURPOSE: Test uploading sketches to editor.p5js.org programmatically
 *
 * REQUIREMENTS:
 * - Node.js 14+
 * - npm install axios
 *
 * SETUP:
 * 1. Log into editor.p5js.org in your browser
 * 2. Open DevTools → Application → Cookies
 * 3. Copy the value of 'connect.sid' cookie
 * 4. Paste it in the SESSION_COOKIE variable below
 *
 * OR (if Personal Access Tokens are working):
 * 1. Generate token in Account Settings
 * 2. Set USERNAME and ACCESS_TOKEN variables
 *
 * USAGE:
 * node test-upload.js
 */

const axios = require('axios');

// ============================================================================
// CONFIGURATION - UPDATE THESE VALUES
// ============================================================================

// Method 1: Session Cookie (copy from browser DevTools)
const SESSION_COOKIE = 'YOUR_SESSION_COOKIE_HERE';

// Method 2: Personal Access Token (if available)
const USERNAME = 'your_username';
const ACCESS_TOKEN = 'your_personal_access_token';

// Which authentication method to use
const USE_TOKEN_AUTH = false; // Set to true to use token, false for cookie

// ============================================================================
// API CLIENT CONFIGURATION
// ============================================================================

const API_BASE_URL = 'https://editor.p5js.org/api';

// Create axios instance with default config
const apiClient = axios.create({
  baseURL: API_BASE_URL,
  timeout: 10000,
  headers: {
    'Content-Type': 'application/json',
    'User-Agent': 'p5-api-test-client/1.0'
  }
});

// Add authentication to requests
apiClient.interceptors.request.use(config => {
  if (USE_TOKEN_AUTH) {
    // Method 1: HTTP Basic Auth with Personal Access Token
    const auth = Buffer.from(`${USERNAME}:${ACCESS_TOKEN}`).toString('base64');
    config.headers.Authorization = `Basic ${auth}`;
    console.log('🔑 Using Token Authentication');
  } else {
    // Method 2: Session Cookie
    config.headers.Cookie = `connect.sid=${SESSION_COOKIE}`;
    config.withCredentials = true;
    console.log('🔑 Using Session Cookie Authentication');
  }
  return config;
});

// Log requests and responses
apiClient.interceptors.request.use(config => {
  console.log(`\n📤 ${config.method.toUpperCase()} ${config.url}`);
  return config;
});

apiClient.interceptors.response.use(
  response => {
    console.log(`✅ ${response.status} ${response.statusText}`);
    return response;
  },
  error => {
    if (error.response) {
      console.log(`❌ ${error.response.status} ${error.response.statusText}`);
      console.log('Response:', error.response.data);
    } else {
      console.log('❌ Error:', error.message);
    }
    return Promise.reject(error);
  }
);

// ============================================================================
// SKETCH TEMPLATES
// ============================================================================

const TEMPLATE_BASIC = {
  name: 'Test Sketch - Basic',
  files: [
    {
      name: 'sketch.js',
      content: `function setup() {
  createCanvas(400, 400);
}

function draw() {
  background(220);

  // Draw a circle that follows the mouse
  fill(255, 0, 0);
  circle(mouseX, mouseY, 50);

  // Display coordinates
  fill(0);
  textAlign(CENTER);
  text(\`x: \${mouseX}, y: \${mouseY}\`, width / 2, 20);
}`,
      fileType: 'file'
    },
    {
      name: 'index.html',
      content: `<!DOCTYPE html>
<html lang="en">
  <head>
    <meta charset="utf-8">
    <meta name="viewport" content="width=device-width, initial-scale=1.0">
    <title>Test Sketch</title>
    <style>
      body {
        padding: 0;
        margin: 0;
      }
    </style>
    <script src="https://cdnjs.cloudflare.com/ajax/libs/p5.js/1.4.0/p5.js"></script>
    <script src="sketch.js"></script>
  </head>
  <body>
  </body>
</html>`,
      fileType: 'file'
    }
  ]
};

const TEMPLATE_ANIMATION = {
  name: 'Test Sketch - Animation',
  files: [
    {
      name: 'sketch.js',
      content: `let angle = 0;

function setup() {
  createCanvas(400, 400);
}

function draw() {
  background(220);

  translate(width / 2, height / 2);
  rotate(angle);

  fill(255, 0, 0);
  rectMode(CENTER);
  rect(0, 0, 100, 100);

  angle += 0.01;
}`,
      fileType: 'file'
    },
    {
      name: 'index.html',
      content: `<!DOCTYPE html>
<html lang="en">
  <head>
    <meta charset="utf-8">
    <meta name="viewport" content="width=device-width, initial-scale=1.0">
    <title>Animation Test</title>
    <style>
      body {
        padding: 0;
        margin: 0;
      }
    </style>
    <script src="https://cdnjs.cloudflare.com/ajax/libs/p5.js/1.4.0/p5.js"></script>
    <script src="sketch.js"></script>
  </head>
  <body>
  </body>
</html>`,
      fileType: 'file'
    }
  ]
};

// ============================================================================
// API FUNCTIONS
// ============================================================================

/**
 * Test authentication by checking access
 */
async function testAuthentication() {
  console.log('\n🔍 Testing Authentication...');
  try {
    // Try the test endpoint if it exists
    const response = await apiClient.get('/auth/access-check');
    console.log('✅ Authentication successful!');
    return true;
  } catch (error) {
    if (error.response?.status === 404) {
      console.log('⚠️  Test endpoint not found (this is okay)');
      return null; // Unknown, endpoint doesn't exist
    }
    console.log('❌ Authentication failed');
    return false;
  }
}

/**
 * Create a new sketch
 */
async function createSketch(sketchData) {
  console.log(`\n📝 Creating sketch: "${sketchData.name}"...`);

  try {
    const response = await apiClient.post('/projects', sketchData);

    const project = response.data;
    console.log('✅ Sketch created successfully!');
    console.log('\n📊 Project Details:');
    console.log('  ID:', project.id);
    console.log('  Name:', project.name);
    console.log('  Owner:', project.owner?.username || project.user?.username);
    console.log('  Created:', project.createdAt);

    const username = project.owner?.username || project.user?.username || USERNAME;
    const sketchUrl = `https://editor.p5js.org/${username}/sketches/${project.id}`;
    console.log('\n🔗 Sketch URL:');
    console.log('  ', sketchUrl);
    console.log('\n🎨 Full View:');
    console.log('  ', `https://editor.p5js.org/${username}/full/${project.id}`);

    return project;
  } catch (error) {
    console.log('❌ Failed to create sketch');
    if (error.response) {
      console.log('Status:', error.response.status);
      console.log('Response:', JSON.stringify(error.response.data, null, 2));
    }
    throw error;
  }
}

/**
 * Get project details
 */
async function getProject(projectId) {
  console.log(`\n🔍 Fetching project ${projectId}...`);

  try {
    const response = await apiClient.get(`/projects/${projectId}`);
    console.log('✅ Project fetched successfully');
    return response.data;
  } catch (error) {
    console.log('❌ Failed to fetch project');
    throw error;
  }
}

/**
 * Update a sketch
 */
async function updateSketch(projectId, updates) {
  console.log(`\n✏️  Updating sketch ${projectId}...`);

  try {
    const response = await apiClient.put(`/projects/${projectId}`, updates);
    console.log('✅ Sketch updated successfully');
    return response.data;
  } catch (error) {
    console.log('❌ Failed to update sketch');
    throw error;
  }
}

/**
 * Delete a sketch
 */
async function deleteSketch(projectId) {
  console.log(`\n🗑️  Deleting sketch ${projectId}...`);

  try {
    await apiClient.delete(`/projects/${projectId}`);
    console.log('✅ Sketch deleted successfully');
    return true;
  } catch (error) {
    console.log('❌ Failed to delete sketch');
    throw error;
  }
}

/**
 * List user's sketches (public endpoint, no auth needed)
 */
async function listUserSketches(username) {
  console.log(`\n📋 Listing sketches for user: ${username}...`);

  try {
    const response = await axios.get(`${API_BASE_URL}/${username}/sketches`);
    console.log(`✅ Found ${response.data.length} sketches`);

    response.data.slice(0, 5).forEach((sketch, i) => {
      console.log(`\n  ${i + 1}. ${sketch.name}`);
      console.log(`     ID: ${sketch.id}`);
      console.log(`     Updated: ${sketch.updatedAt}`);
    });

    if (response.data.length > 5) {
      console.log(`\n  ... and ${response.data.length - 5} more`);
    }

    return response.data;
  } catch (error) {
    console.log('❌ Failed to list sketches');
    throw error;
  }
}

// ============================================================================
// TEST SCENARIOS
// ============================================================================

/**
 * Run all tests
 */
async function runTests() {
  console.log('╔════════════════════════════════════════════════════════════════╗');
  console.log('║    p5.js Web Editor API - Upload Test Suite                   ║');
  console.log('╚════════════════════════════════════════════════════════════════╝');

  try {
    // Test 1: Authentication
    await testAuthentication();

    // Test 2: Create a basic sketch
    const project = await createSketch(TEMPLATE_BASIC);

    // Test 3: Fetch the created project
    await getProject(project.id);

    // Test 4: Update the sketch
    const updates = {
      name: 'Test Sketch - Updated',
      files: TEMPLATE_ANIMATION.files
    };
    await updateSketch(project.id, updates);

    // Test 5: List user's sketches
    if (USERNAME !== 'your_username') {
      await listUserSketches(USERNAME);
    }

    // Test 6: Clean up (optional - comment out if you want to keep the sketch)
    // await deleteSketch(project.id);

    console.log('\n╔════════════════════════════════════════════════════════════════╗');
    console.log('║    ✅ All tests completed successfully!                        ║');
    console.log('╚════════════════════════════════════════════════════════════════╝');

  } catch (error) {
    console.log('\n╔════════════════════════════════════════════════════════════════╗');
    console.log('║    ❌ Tests failed                                             ║');
    console.log('╚════════════════════════════════════════════════════════════════╝');
    console.error('\nError details:', error.message);
    process.exit(1);
  }
}

// ============================================================================
// VALIDATION
// ============================================================================

function validateConfig() {
  console.log('🔧 Validating configuration...');

  if (!USE_TOKEN_AUTH && SESSION_COOKIE === 'YOUR_SESSION_COOKIE_HERE') {
    console.error('\n❌ ERROR: SESSION_COOKIE not configured!');
    console.error('\n📖 Instructions:');
    console.error('1. Log into editor.p5js.org in your browser');
    console.error('2. Open DevTools (F12) → Application → Cookies');
    console.error('3. Copy the value of "connect.sid" cookie');
    console.error('4. Paste it in this script as SESSION_COOKIE');
    console.error('\nOR set USE_TOKEN_AUTH = true and configure token credentials\n');
    process.exit(1);
  }

  if (USE_TOKEN_AUTH && (USERNAME === 'your_username' || ACCESS_TOKEN === 'your_personal_access_token')) {
    console.error('\n❌ ERROR: Token credentials not configured!');
    console.error('\n📖 Instructions:');
    console.error('1. Log into editor.p5js.org');
    console.error('2. Go to Account Settings');
    console.error('3. Generate a Personal Access Token');
    console.error('4. Set USERNAME and ACCESS_TOKEN in this script');
    console.error('\nOR set USE_TOKEN_AUTH = false and use session cookie instead\n');
    process.exit(1);
  }

  console.log('✅ Configuration valid\n');
}

// ============================================================================
// MAIN
// ============================================================================

if (require.main === module) {
  // Check if axios is installed
  try {
    require.resolve('axios');
  } catch (e) {
    console.error('❌ axios is not installed!');
    console.error('Run: npm install axios');
    process.exit(1);
  }

  validateConfig();
  runTests().catch(error => {
    console.error('Fatal error:', error);
    process.exit(1);
  });
}

// Export functions for use as module
module.exports = {
  createSketch,
  getProject,
  updateSketch,
  deleteSketch,
  listUserSketches
};
